contracts/SnappAuction.sol

```diff
@@ -252,35 +254,46 @@ contract SnappAuction is SnappBase {
         uint slot,
         bytes32 _currStateRoot,
-        bytes32 _newStateRoot,
+        bytes32 newStateRoot,                 // Only needed the case of fallback
         bytes32 _orderHash,
         uint128[] memory _standingOrderIndex,
-        bytes memory pricesAndVolumes
-    )
-        public onlyOwner()
-    {
+        bytes memory pricesAndVolumes         // Can be empty in the trivial case.
+    ) public onlyOwner() {
+        // Auction related constraints (slot exists, is inactive, no previous auction pending and not already applied)
         require(slot != MAX_UINT && slot <= auctionIndex, "Requested auction slot does not exist");
         require(slot == 0 || auctions[slot-1].appliedAccountStateIndex != 0, "Must apply auction slots in order!");
         require(auctions[slot].appliedAccountStateIndex == 0, "Auction already applied");
-        require(
-            calculateOrderHash(slot, _standingOrderIndex) == _orderHash,
-            "Order hash doesn't agree"
-        );
+
+        // State related constraints (order hash and state root agree)
+        require(coreData.stateRoots[stateIndex()] == _currStateRoot, "Incorrect state root");
+        require(calculateOrderHash(slot, _standingOrderIndex) == _orderHash, "Order hash doesn't agree");
+
+        // Phase related constraints
         require(
             block.timestamp > auctions[slot].creationTimestamp + 3 minutes ||
                 auctions[slot].numOrders == maxUnreservedOrderCount(),
             "Requested auction slot is still active"
         );
-        require(coreData.stateRoots[stateIndex()] == _currStateRoot, "Incorrect state root");
-
-        coreData.stateRoots.push(_newStateRoot);
-        auctions[slot].appliedAccountStateIndex = stateIndex();
-
-        // Store solution information in shaHash of pendingBatch (required for snark proof)
-        auctions[slot].solutionHash = sha256(pricesAndVolumes);
-        auctions[slot].solutionAcceptedTime = block.timestamp;
+        require(
+            block.timestamp > biddingStartTime(slot) + 3 minutes,
+            "Requested auction still in bidding phase or earlier"
+        );
 
-        emit AuctionSettlement(slot, stateIndex(), _newStateRoot, pricesAndVolumes);
+        if (block.timestamp < biddingStartTime(slot) + 270 seconds && auctions[slot].solver != address(0)) {
+            // Winner Apply Auction
+            require(
+                auctions[slot].solver == msg.sender,
+                "Only winner of bidding phase may apply auction here"
+            );
+            internalApplyAuction(slot, auctions[slot].tentativeState, pricesAndVolumes);
+        } else if (block.timestamp < biddingStartTime(slot) + 6 minutes) {
+            // Fallback Apply Auction
+            internalApplyAuction(slot, newStateRoot, pricesAndVolumes);
+        } else {
+            // Trivial Apply Auction
+            bytes memory trivialSolution;  // p_i = 1 and (bA, sA)_j = (0, 0) \forall i, j
```

Comment:
What is the value of this? If it's empty can we make it explicit `0x0`?

Reply:
It is returned as null. You should be able to see the assertion in the uni test when compared to the trivial solution application. I guess we could make it explicit...

Reply:
I have tried `0x0`, but got something like this 

```
TypeError: Type int_const 0 is not implicitly convertible to expected type bytes memory.
            bytes memory trivialSolution = 0x0
```

'm ok with null for now if everyone else is.